WalletWasabi/Blockchain/Analysis/BlockchainAnalyzer.cs

```diff
@@ -3,60 +3,182 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Text;
+using WalletWasabi.Blockchain.Keys;
 using WalletWasabi.Blockchain.Transactions;
 
 namespace WalletWasabi.Blockchain.Analysis
 {
 	public class BlockchainAnalyzer
 	{
-		public BlockchainAnalyzer(int privacyLevelThreshold, Money dustThreshold)
+		public BlockchainAnalyzer(int privacyLevelThreshold)
 		{
 			PrivacyLevelThreshold = privacyLevelThreshold;
-			DustThreshold = dustThreshold;
 		}
 
 		public int PrivacyLevelThreshold { get; }
-		public Money DustThreshold { get; }
 
 		/// <summary>
 		/// Sets clusters and anonymity sets for related HD public keys.
 		/// </summary>
 		public void Analyze(SmartTransaction tx)
 		{
-			// Dust attack could ruin the analysis, so it's better not to incorporate this.
-			// For example if someone would dust a mixed coin of ours, its anon score would end up being 1,
-			// even though the duster does not know who it sent the money.
-			// FTR at the time of writing this comment the transaction processor doesn't even let dust to come this far,
-			// but it's better to be prepared for future changes if it's not too computationally expensive.
-			foreach (var newCoin in tx.WalletOutputs.Where(x => x.Amount > DustThreshold))
+			var inputCount = tx.Transaction.Inputs.Count;
+			var outputCount = tx.Transaction.Outputs.Count;
+
+			var ownInputCount = tx.WalletInputs.Count;
+			var ownOutputCount = tx.WalletOutputs.Count;
+
+			if (inputCount == ownInputCount)
+			{
+				if (outputCount == ownOutputCount)
+				{
+					AnalyzeSelfSpend(tx);
+				}
+				else
+				{
+					AnalyzeNormalSpend(tx);
+				}
+			}
+			else
+			{
+				AnalyzeCoinjoin(tx);
+			}
+
+			AnalyzeClusters(tx);
+		}
+
+		private void AnalyzeCoinjoin(SmartTransaction tx)
+		{
+			AnalyzeWalletInputs(tx, out HashSet<HdPubKey> distinctWalletInputPubKeys, out int inheritedAnonset);
+
+			var indistinguishableWalletOutputs = tx
+				.WalletOutputs.GroupBy(x => x.Amount)
+				.ToDictionary(x => x.Key, y => y.Count());
+
+			foreach (var newCoin in tx.WalletOutputs.ToArray())
 			{
-				// Calculate anonymity sets.
 				// Get the anonymity set of i-th output in the transaction.
-				var anonset = tx.Transaction.GetAnonymitySet(newCoin.Index);
+				var anonset = inheritedAnonset;
+
+				// Calculating gained anonimity shall be limited by the number of inputs.
+				// Although one person may create many equal outputs and it would seem like we successfully get lost
+				// within this elephant, it's not the case as the elephant is more likely to ruin its own privacy, so let's not inflate anonsets for this.
+				anonset += Math.Min(tx.Transaction.Inputs.Count, tx.Transaction.GetAnonymitySet(newCoin.Index));
```

Comment:
GetAnonymitySet calculates with the input number already

Reply:
Good catch! [`79efe2f` (#4724)](https://github.com/zkSNACKs/WalletWasabi/pull/4724/commits/79efe2f6c98c3a92bf067efd46af28b64637b010)